File: app/Python/qr_extractor.py
#!/usr/bin/env python3
"""
Script para extraer códigos QR de archivos PDF
Requiere: pip install PyMuPDF pyzbar pillow opencv-python
"""

import sys
import json
import fitz  # PyMuPDF
from PIL import Image
from pyzbar.pyzbar import decode
import cv2
import numpy as np
import io
import os

def extract_qr_url_from_pdf(pdf_path):
    """
    Extrae URL de código QR de un archivo PDF
    Versión optimizada y simplificada
    """
    try:
        # Verificar que el archivo existe
        if not os.path.exists(pdf_path):
            return None
        
        # Abrir el PDF
        doc = fitz.open(pdf_path)
        
        for page_num in range(len(doc)):
            page = doc.load_page(page_num)
            
            # Convertir página a imagen con mejor resolución
            pix = page.get_pixmap()
            img_data = pix.tobytes()
            
            # Convertir a PIL Image
            img = Image.open(io.BytesIO(img_data))
            
            # Convertir a formato compatible con OpenCV
            img_cv = cv2.cvtColor(np.array(img), cv2.COLOR_RGB2BGR)
            
            # Buscar QR con pyzbar (más rápido y confiable)
            decoded_objs = decode(img_cv)
            
            for obj in decoded_objs:
                data = obj.data.decode('utf-8')
                if data.startswith('http'):
                    doc.close()
                    return data
        
        doc.close()
        return None
        
    except Exception as e:
        return None

def extract_qr_from_pdf(pdf_path):
    """
    Función wrapper para mantener compatibilidad con el controlador PHP
    """
    try:
        url = extract_qr_url_from_pdf(pdf_path)
        
        if url:
            return {"success": True, "url": url}
        else:
            return {"success": False, "error": "No se encontró código QR válido en el PDF"}
            
    except Exception as e:
        return {"success": False, "error": f"Error procesando PDF: {str(e)}"}

def main():
    if len(sys.argv) != 2:
        print(json.dumps({"success": False, "error": "Uso: python qr_extractor.py <ruta_pdf>"}))
        sys.exit(1)
    
    pdf_path = sys.argv[1]
    result = extract_qr_from_pdf(pdf_path)
    print(json.dumps(result))

if __name__ == "__main__":
    main()

File: app/JavaScript/qr_extractor.js
#!/usr/bin/env node
/**
 * Script para extraer códigos QR de archivos PDF usando Node.js
 * Requiere: npm install pdf-poppler jimp qrcode-reader
 */

const fs = require('fs');
const path = require('path');
const pdf = require('pdf-poppler');
const Jimp = require('jimp');
const QrCode = require('qrcode-reader');

async function extractQrFromPdf(pdfPath) {
    try {
        // Verificar que el archivo existe
        if (!fs.existsSync(pdfPath)) {
            return { success: false, error: "Archivo PDF no encontrado" };
        }

        // Configuración para convertir PDF a imágenes
        const options = {
            format: 'png',
            out_dir: path.dirname(pdfPath),
            out_prefix: 'temp_page',
            page: null // Convertir todas las páginas
        };

        // Convertir PDF a imágenes
        const pages = await pdf.convert(pdfPath, options);
        
        // Procesar cada página
        for (let i = 1; i <= pages.length; i++) {
            const imagePath = path.join(options.out_dir, `${options.out_prefix}-${i}.png`);
            
            try {
                // Leer imagen con Jimp
                const image = await Jimp.read(imagePath);
                
                // Crear instancia del lector QR
                const qr = new QrCode();
                
                // Promesa para leer QR
                const qrResult = await new Promise((resolve, reject) => {
                    qr.callback = (err, value) => {
                        if (err) {
                            reject(err);
                        } else {
                            resolve(value);
                        }
                    };
                    qr.decode(image.bitmap);
                });
                
                // Verificar si es una URL válida
                if (qrResult && qrResult.result && qrResult.result.startsWith('http')) {
                    // Limpiar archivos temporales
                    cleanupTempFiles(options.out_dir, options.out_prefix, pages.length);
                    return { success: true, url: qrResult.result };
                }
                
            } catch (pageError) {
                // Continuar con la siguiente página
                continue;
            } finally {
                // Limpiar archivo de página actual
                if (fs.existsSync(imagePath)) {
                    fs.unlinkSync(imagePath);
                }
            }
        }
        
        // Limpiar archivos temporales restantes
        cleanupTempFiles(options.out_dir, options.out_prefix, pages.length);
        
        return { success: false, error: "No se encontró código QR válido en el PDF" };
        
    } catch (error) {
        return { success: false, error: `Error procesando PDF: ${error.message}` };
    }
}

function cleanupTempFiles(dir, prefix, pageCount) {
    for (let i = 1; i <= pageCount; i++) {
        const filePath = path.join(dir, `${prefix}-${i}.png`);
        if (fs.existsSync(filePath)) {
            try {
                fs.unlinkSync(filePath);
            } catch (e) {
                // Ignorar errores de limpieza
            }
        }
    }
}

async function main() {
    if (process.argv.length !== 3) {
        console.log(JSON.stringify({ success: false, error: "Uso: node qr_extractor.js <ruta_pdf>" }));
        process.exit(1);
    }
    
    const pdfPath = process.argv[2];
    const result = await extractQrFromPdf(pdfPath);
    console.log(JSON.stringify(result));
}

if (require.main === module) {
    main().catch(error => {
        console.log(JSON.stringify({ success: false, error: error.message }));
        process.exit(1);
    });
}